File: android-programming-3e/03_activity_lifecycle/java/GeoQuiz Challenge/app/src/main/java/com/bignerdranch/geoquiz/MainActivity.java
package com.bignerdranch.geoquiz;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import java.util.HashMap;
import java.util.Map;

public class MainActivity extends AppCompatActivity {
    private Button mTrueButton;
    private Button mFalseButton;
    private ImageButton mPrevButton;
    private ImageButton mNextButton;
    private TextView mQuestionTextView;

    private Question[] mQuestionBank = new Question[] {
        new Question(R.string.question_australia, true),
        new Question(R.string.question_oceans, true),
        new Question(R.string.question_mideast, false),
        new Question(R.string.question_africa, false),
        new Question(R.string.question_americas, true),
        new Question(R.string.question_asia, true)
    };

    private int mCurrentIndex = 0;
    private String TAG = "QuizActivity";
    private String KEY_INDEX = "index";
    private HashMap<Integer, Boolean> mAnswerBank = new HashMap<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mQuestionTextView = (TextView) findViewById(R.id.question_text_view);
        mTrueButton = (Button) findViewById(R.id.true_button);
        mFalseButton = (Button) findViewById(R.id.false_button);
        mPrevButton = (ImageButton) findViewById(R.id.prev_button);
        mNextButton = (ImageButton) findViewById(R.id.next_button);
        mTrueButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                checkAnswer(true);
            }
        });

        mFalseButton.setOnClickListener(v -> checkAnswer(false));

        mPrevButton.setOnClickListener(v -> {
            --mCurrentIndex;
            if (mCurrentIndex < 0) mCurrentIndex = mQuestionBank.length - 1;
            updateQuestion();
        });

        mNextButton.setOnClickListener(v -> {
            mCurrentIndex = (mCurrentIndex + 1) % mQuestionBank.length;
            updateQuestion();
        });

        mQuestionTextView.setOnClickListener(v -> {
            mCurrentIndex = (mCurrentIndex + 1) % mQuestionBank.length;
            updateQuestion();
        });

        if (savedInstanceState != null) {
            mCurrentIndex = savedInstanceState.getInt(KEY_INDEX);
        }

        updateQuestion();
    }

    @Override
    protected void onStart() {
        super.onStart();
        Log.d(TAG, "onStart() called");
    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.d(TAG, "onResume() called");
    }

    @Override
    protected void onPause() {
        super.onPause();
        Log.d(TAG, "onPause() called");
    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.d(TAG, "onStop() called");
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.d(TAG, "onDestroy() called");
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        Log.d(TAG, "onSaveInstanceState() called");
        outState.putInt(KEY_INDEX, mCurrentIndex);
    }


    private void updateQuestion() {
        int question = mQuestionBank[mCurrentIndex].getTextResId();
        mQuestionTextView.setText(question);
        boolean isEnabled = !mAnswerBank.containsKey(mCurrentIndex);
        mTrueButton.setEnabled(isEnabled);
        mFalseButton.setEnabled(isEnabled);
    }

    private void checkAnswer(boolean userPressedTrue) {
        boolean answerIsTrue = mQuestionBank[mCurrentIndex].isAnswerTrue();
        int messageResId = 0;
        if (userPressedTrue == answerIsTrue) {
            messageResId = R.string.correct_toast;
            mAnswerBank.put(mCurrentIndex, true);
        } else {
            messageResId = R.string.incorrect_toast;
            mAnswerBank.put(mCurrentIndex, false);
        }

        Toast.makeText(this, messageResId, Toast.LENGTH_SHORT).show();
        updateQuestion();
        checkScore();
    }

    private void checkScore() {
        if (mAnswerBank.size() == mQuestionBank.length) {
            int correctAnswers = 0;
            for (Map.Entry<Integer, Boolean> i: mAnswerBank.entrySet()) {
                if (i.getValue()) {
                    correctAnswers++;
                }
            }

            int score = (int)(Math.round((double)(correctAnswers) / mAnswerBank.size() * 100));
            String scoreText = getApplicationContext().getResources().getString(
                    R.string.quiz_score, score);
            Toast.makeText(this, scoreText, Toast.LENGTH_LONG).show();
        }
    }
}

File: android-programming-3e/03_activity_lifecycle/kotlin/GeoQuiz Challenge/app/src/main/java/com/bignerdranch/geoquiz/MainActivity.kt
package com.bignerdranch.geoquiz

import android.os.Bundle
import android.util.Log
import android.view.Gravity
import android.widget.Button
import android.widget.ImageButton
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import kotlin.concurrent.thread

class MainActivity : AppCompatActivity() {
    private lateinit var trueBtn: Button
    private lateinit var falseBtn: Button
    private lateinit var prevBtn: ImageButton
    private lateinit var nextBtn: ImageButton
    private lateinit var questionTextView: TextView
    private val questionBank = listOf(
        Question(R.string.question_australia, true),
        Question(R.string.question_oceans, true),
        Question(R.string.question_mideast, false),
        Question(R.string.question_africa, false),
        Question(R.string.question_americas, true),
        Question(R.string.question_asia, true)
    )

    private val answerBank = HashMap<Int, Boolean>()
    private var currentIndex = 0
    private val TAG = "QuizActivity"
    private val KEY_INDEX = "index"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.d(TAG, "onCreate(Bundle) called")
        setContentView(R.layout.activity_main)
        questionTextView = findViewById(R.id.question_text_view)
        trueBtn = findViewById(R.id.true_button)
        falseBtn = findViewById(R.id.false_button)
        prevBtn = findViewById(R.id.prev_button)
        nextBtn = findViewById(R.id.next_button)
        trueBtn.setOnClickListener {
            checkAnswer(true)
        }

        falseBtn.setOnClickListener {
            checkAnswer(false)
        }

        prevBtn.setOnClickListener {
            --currentIndex
            if (currentIndex < 0) {
                currentIndex = questionBank.size - 1
            }

            updateQuestion()
        }

        nextBtn.setOnClickListener {
            currentIndex = (currentIndex + 1) % questionBank.size
            updateQuestion()
        }

        questionTextView.setOnClickListener {
            currentIndex = (currentIndex + 1) % questionBank.size
            updateQuestion()
        }

        savedInstanceState?.let {
            currentIndex = it.getInt(KEY_INDEX)
        }

        updateQuestion()
    }

    override fun onStart() {
        super.onStart()
        Log.d(TAG, "onStart() called")
    }

    override fun onResume() {
        super.onResume()
        Log.d(TAG, "onResume() called")
    }

    override fun onPause() {
        super.onPause()
        Log.d(TAG, "onPause() called")
    }

    override fun onStop() {
        super.onStop()
        Log.d(TAG, "onStop() called")
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d(TAG, "onDestroy() called")
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        Log.d(TAG, "onSaveInstanceState() called")
        outState.putInt(KEY_INDEX, currentIndex)
    }

    private fun updateQuestion() {
        questionTextView.setText(questionBank[currentIndex].textResId)
        val isEnabled = !answerBank.containsKey(currentIndex)
        trueBtn.isEnabled = isEnabled
        falseBtn.isEnabled = isEnabled
    }

    private fun checkAnswer(userPressedTrue: Boolean) {
        val answerIsTrue = questionBank[currentIndex].answerTrue
        val messageResId = if (userPressedTrue == answerIsTrue) {
            answerBank[currentIndex] = true
            R.string.correct_toast
        } else {
            answerBank[currentIndex] = false
            R.string.incorrect_toast
        }

        Toast.makeText(this, messageResId, Toast.LENGTH_SHORT).show()
        updateQuestion()
        checkScore()
    }

    private fun checkScore() {
        if (answerBank.size == questionBank.size) {
            var correctAnswers: Int = 0
            for ((_, isCorrect) in answerBank) {
                if (isCorrect) {
                    correctAnswers++
                }
            }
            val score = Math.round(correctAnswers.toDouble() / answerBank.size * 100).toInt()
            val scoreText = applicationContext.resources.getString(R.string.quiz_score, score)
            Toast.makeText(this, scoreText, Toast.LENGTH_LONG).show()
        }
    }
}